lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,234 @@
+        is not parsable, the json error will bubble up.
+        """
+        block_dicts = json.loads(self._blocks_json)
+        return [
+            BlockRecord(data['weight'], data['max_score'], data['block_key'])
+            for data in block_dicts
+        ]
+
+    @blocks.setter
+    def blocks(self, value):
+        """
+        Not implemented, as VisibleBlocks instances are intended to be write-once and not change after creation.
+        """
+        raise NotImplementedError(
+            "Property 'blocks' cannot be modified on an existing VisibleBlocks model. Create a new instance."
+        )
+
+
+class BlockRecord(object):
+    """
+    An object encapsulating all relevant information for representing a block at the time it was used for grade
+    calculation.
+    """
+
+    def __init__(self, weight, max_score, locator):
+        """
+        Creates a BlockRecord.
+
+        Params:
+            weight (float)
+            max_score (float)
+            locator (BlockUsageLocator or string)
+        """
+        self.weight = weight
+        self.max_score = max_score
+        # pylint: disable=protected_access
+        self.block_key = locator._to_string() if isinstance(locator, BlockUsageLocator) else locator
+
+    def to_dict(self):
+        """
+        Serialize this object to a dict object.
+        """
+        return {
+            'weight': self.weight,
+            'max_score': self.max_score,
+            'block_key': self.block_key,
+        }
+
+    @property
+    def locator(self):
+        """
+        Convenience property allowing a BlockUsageLocator to be returned from this object's serialized data.
+        """
+        return BlockUsageLocator._from_string(self.block_key)  # pylint: disable=protected_access
+
+
+class PersistentSubsectionGrade(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+
+    TODO: here are the other query patterns listed in the ticket that are not currently used. Should any of these
+    indices be built?
+        user_id, course_id, content_type, is_valid
+        course_id, edit-timestamp
+        edit-timestamp
+    """
+
+    class Meta(object):
+        index_together = [
+            ('user_id', 'usage_key')
+        ]
+
+        unique_together = (('user_id', 'usage_key'))
+
+    id = UnsignedBigIntAutoField(primary_key=True)  # pylint: disable=invalid-name
+    subtree_edited_date = models.DateTimeField('last content edit timestamp')
+    user_id = models.CharField(max_length=255)
+    earned_all = models.IntegerField()
+    possible_all = models.IntegerField()
+    earned_graded = models.IntegerField()
+    possible_graded = models.IntegerField()
+
+    course_id = CourseKeyField(max_length=255)
+    usage_key = UsageKeyField(max_length=255)
+    course_version = models.CharField('guid of latest course version', max_length=255)
+
+    #is_valid = models.BinaryField()  # Might be needed if doing async updates
```

Comment:
I'd lean towards removing it altogether right now, and adding it only if needed in the follow-up ticket.  Then if we don't need to do async, the code doesn't need to be touched.